### DISCUSSION TITLE
[Question][Hw5] Question about an unreachable code

### DISCUSSION THREAD

**Discussion by askme143** [prosyslab-classroom/cs348-information-security#595] [Q&A]
We know that a division by a variable could be converted into a division by a constant by llvm for some reasons. We also know that this can make the "Division by 0" problem more trickier.

For example,

```C
// C code
#include "homework.h"

int main() {
  int x = 1;
  int temp = 0;
  x = 0;
  if (temp) {
    10 / x; // unreachable                          <------------ Division by a variable 0, but unreachable
  }
  return 0;
}
```

```llvm
; LLVM
(...)
define dso_local i32 @main() #0 !dbg !7 {
entry:
  call void @llvm.dbg.value(metadata i32 1, metadata !11, metadata !DIExpression()), !dbg !12
  call void @llvm.dbg.value(metadata i32 0, metadata !13, metadata !DIExpression()), !dbg !12
  call void @llvm.dbg.value(metadata i32 0, metadata !11, metadata !DIExpression()), !dbg !12
  %tobool = icmp ne i32 0, 0, !dbg !14
  br i1 %tobool, label %if.then, label %if.end, !dbg !16

if.then:                                          ; preds = %entry
  %div = sdiv i32 10, 0, !dbg !17                 ;<------------ Division by a constant 0
  br label %if.end, !dbg !19

if.end:                                           ; preds = %if.then, %entry
  ret i32 0, !dbg !20
}
(...)
```

Line 8 is an unreachable line. However, since the checker in our skeleton checks "division by 0" by evaluating each node, it will print a false alarm. Should I consider cases like this?

**Comment by KihongHeo** [comment]
That is because LLVM optimized the code.
In principle, that kind of false alarms can be handled with the sign domain. But the current skeleton code is not intended to support that.

You can ignore such case. I will ask TAs not to have such test cases this class.
Good suggestion for next year students!
